File: python-hypothesis/src/poker.py
from enum import Enum
import random
from typing import List

class Suit(Enum):
    SPADES = "Spades"
    DIAMONDS = "Diamonds"
    CLUBS = "Clubs"
    HEARTS = "Hearts"

class Rank(Enum):
    TWO = "2"
    THREE = "3"
    FOUR = "4"
    FIVE = "5"
    SIX = "6"
    SEVEN = "7"
    EIGHT = "8"
    NINE = "9"
    TEN = "10"
    JACK = "Jack"
    QUEEN = "Queen"
    KING = "King"
    ACE = "Ace"

class Card:
    def __init__(self, suit: Suit, rank: Rank):
        self.suit = suit
        self.rank = rank

    def __repr__(self):
        return f"{self.rank.value} of {self.suit.value}"
    
    def __eq__(self, other):
        if not isinstance(other, Card):
            return NotImplemented
        return self.suit == other.suit and self.rank == other.rank

    def __hash__(self):
        return hash((self.suit, self.rank))

class Deck:
    def __init__(self):
        self.cards : List[List[Card]] = [Card(suit, rank) for suit in Suit for rank in Rank]
        self.shuffle()
        
    def __eq__(self, other):
        if not isinstance(other, Deck):
            return NotImplemented
        return set(self.cards) == set(other.cards)

    def shuffle(self):
        random.shuffle(self.cards)

    def deal(self, num_players: int) -> List[List[Card]]:
        if num_players < 1:
            raise ValueError("Must have at least one player to play!")
        if num_players * 5 > len(self.cards):
            raise ValueError("Not enough cards to deal!")
        
        hands : List[List[Card]] = []
        
        # Deal 5 cards to each player
        for _ in range(num_players):
            hand = self.cards[:5]
            hands.append(hand)
            self.cards = self.cards[5:]
        return hands

File: rust-quickcheck/src/poker.rs
#[cfg(test)]
mod tests {
    use quickcheck_macros::quickcheck;

    #[derive(Clone, Debug)]
    enum Suit {
        Spades,
        Diamonds,
        Clubs,
        Hearts
    }

    #[derive(Clone, Debug)]
    struct Card {
        suit: Suit,
        rank: u32 // 1 = ace, 11 = jack, 12 = queen, 13 = king
    }

    fn deal(players: i32) -> Result<Vec<Vec<Card>>, String> {
        if players < 1 { return Err(String::from("Not enough players")) }
        if players > 10 { return Err(String::from("Too many players")) }

        let mut cards: Vec<Card> = Vec::with_capacity(52);
        for i in 1..=13 {
            cards.push(Card { suit: Suit::Spades, rank: i });
            cards.push(Card { suit: Suit::Diamonds, rank: i });
            cards.push(Card { suit: Suit::Clubs, rank: i });
            cards.push(Card { suit: Suit::Hearts, rank: i });
        }

        let mut result: Vec<Vec<Card>> = Vec::with_capacity(players as usize);
        for i in 0..players as usize {
            result.push(cards[i*5 .. (i+1)*5].to_vec())
        }
        Ok(result)
    }

    #[quickcheck]
    fn less_than_one_error(players: i32) -> bool {
        if players >= 1 { return true }
        if let Err(_) = deal(players) { true } else { false }
    }

    #[quickcheck]
    fn more_than_ten_error(players: i32) -> bool {
        if players <= 10 { return true }
        if let Err(_) = deal(players) { true } else { false }
    }

    #[quickcheck]
    fn correct_number_of_results(players: i32) -> bool {
        if players < 1 || players > 10 { return true }
        let result = deal(players).unwrap();
        if result.len() as i32 != players { return false }
        for hand in result { if hand.len() != 5 { return false } }
        true
    }
}
